target/linux/ipq40xx/base-files/etc/init.d/bootcount

```diff
@@ -10,5 +10,8 @@ start() {
 		[ -n "$(fw_printenv bootcount changed 2>/dev/null)" ] &&\
 			echo -e "bootcount\nchanged\n" | /usr/sbin/fw_setenv -s -
 		;;
+	linksys,ea8300)
+		mtd resetbc s_env || true
```

Comment:
Hm, target/linux/ipq40xx/base-files/etc/init.d/zlinksys_recovery also does a ```mtd resetbc s_env```.

@jeffsf @pepe2k @NoTengoBattery 
Wouldn't it be a good idea to merge target/linux/ipq40xx/base-files/etc/init.d/zlinksys_recovery and
target/linux/ipq40xx/base-files/etc/init.d/bootcount ?

Reply:
@chunkeey 

> Wouldn't it be a good idea to merge target/linux/ipq40xx/base-files/etc/init.d/zlinksys_recovery and
> target/linux/ipq40xx/base-files/etc/init.d/bootcount ?

ACK, all of that should be in `bootcount`.
Also, I would remove all echos and comments from the code.

Reply:
If you want to merge the files... Please read the note in `zlinksys_recovery`. The file is called that way on purpose, as the purpose is to run the script at the very end of the boot process.

Reply:
Because...?

Reply:
Do whatever you want.

Reply:
`bootcount` also executes with `START=99` so I don't understand your previous comment. Can't see anything special inside `zlinksys_recovery` which would prevent merging its code into `bootcount`.

Reply:
I think what @NoTengoBattery is refering to the comment in zlinksys_recovery there.

```
# Why? Because the custom sysupgrade script for the device will turn on
# auto_recovery to "yes". And it's the job of this script to set the
# boot boot_count to 0 and then disable auto_recovery, as that condition
# means that the stable release went well.
```

He wants to say that the recovery step has to be the last one because if some other S99* fails after the script has run, the device might still be a "dead duck". So only if the zlinksys_recovery is truly the last thing to run, the firmware recovery mechanism as a whole would work as intended.

Reply:
At least from what I have seen in the EA8300 and based on [some discussion on the EA6350v3 thread](https://forum.openwrt.org/t/add-support-for-linksys-ea6350-v3/18907/288?u=jeff
):

* The factory U-Boot appears to only use the boot counter partition to detect boot loops, not `boot_count` (I no longer set this on the EA8300, as it unnecessarily writes the U-Boot environment on every boot.)
* Disabling `auto_recovery` is probably a bad idea in any situation as the whole point is that if something is wrong with one firmware ("oops, I wrote over the kernel"), the other boots to allow more graceful recovery.

EA6350v3 (from https://openwrt.org/toh/linksys/linksys_ea6350_v3)
```
bootcmd=if test $boot_part = 1; then run bootpart1; else run bootpart2; fi
```
EA8300
```
bootcmd=if test $auto_recovery = no; then bootipq; elif test $boot_part = 1; then run bootpart1; else run bootpart2; fi
```
(note that `bootipq` did not seem functional on my EA8300; perhaps another bit of development cruft)

If there was a problem with leaving `auto_recovery = yes`, I would expect that one of the EA6350v3 users would have long ago mentioned it.

On the WIP responding to the very helpful comments on this PR, I have moved the setting of `auto_recovery` out of the every-boot process and into the sysupgrade process where, in my opinion, it belongs. This way the U-Boot environment partition is not written on every boot, saving flash wear. It is set when it is most likely to be needed, just before a potentially "bad" flash. As `boot_part` is being written at this time (and/or `bootcmd`, depending on board), there is no additional flash required (as all changes to the U-Boot environment are now done in a single call to `fw_setenv`).